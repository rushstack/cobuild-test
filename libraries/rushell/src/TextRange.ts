// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Text coordinates represented as a line number and column number.
 *
 * @remarks
 * The first character in a file is considered to be in column 1 of line 1.
 * The location with column 0 and line 0 is used to represent an empty, unspecified,
 * or unknown location.
 */
export interface ITextLocation {
  line: number;
  column: number;
}

/**
 * Efficiently references a range of text from a string buffer.
 */
export class TextRange {
  /**
   * Used to represent an empty or unknown range.
   */
  public static readonly empty: TextRange = new TextRange('', 0, 0);

  /**
   * The starting index into the associated text buffer.
   *
   * @remarks
   * The text range corresponds to the `range.buffer.substring(range.pos, range.end)`.
   */
  public readonly pos: number;

  /**
   * The (non-inclusive) ending index for the associated text buffer.
   *
   * @remarks
   * The text range corresponds to the `range.buffer.substring(range.pos, range.end)`.
   */
  public readonly end: number;

  /**
   * The string buffer that the `pos` and `end` indexes refer to.
   */
  public readonly buffer: string;

  private constructor(buffer: string, pos: number, end: number) {
    this.buffer = buffer;
    this.pos = pos;
    this.end = end;
    this._validateBounds();
  }

  /**
   * Constructs a TextRange that corresponds to an entire string object.
   */
  public static fromString(buffer: string): TextRange {
    return new TextRange(buffer, 0, buffer.length);
  }

  /**
   * Constructs a TextRange that corresponds to an entire string object.
   */
  public static fromStringRange(buffer: string, pos: number, end: number): TextRange {
    return new TextRange(buffer, pos, end);
  }

  /**
   * Constructs a TextRange that corresponds to a different range of an existing buffer.
   */
  public getNewRange(pos: number, end: number): TextRange {
    return new TextRange(this.buffer, pos, end);
  }

  public isEmpty(): boolean {
    return this.pos === this.end;
  }

  /**
   * Returns the smallest TextRange object that encompasses both ranges.  If there is a gap
   * between the two ranges, it will be included in the encompassing range.
   */
  public getEncompassingRange(other: TextRange): TextRange {
    let newBuffer: string = this.buffer;

    // Allow combining TextRange.empty with a TextRange from a different buffer
    if (other.buffer.length > 0) {
      newBuffer = other.buffer;

      if (this.buffer.length > 0) {
        if (this.buffer !== other.buffer) {
          throw new Error('The ranges cannot be combined because they come from different buffers');
        }
      }
    }

    let newPos: number = this.pos;
    let newEnd: number = this.end;

    if (!other.isEmpty()) {
      if (this.isEmpty()) {
        newPos = other.pos;
        newEnd = other.end;
      } else {
        // Neither range is empty, so combine them
        newPos = Math.min(other.pos, this.pos);
        newEnd = Math.max(other.end, this.end);
      }
    }

    return new TextRange(newBuffer, newPos, newEnd);
  }

  /**
   * Returns the range from the associated string buffer.
   */
  public toString(): string {
    return this.buffer.substring(this.pos, this.end);
  }

  /**
   * Calculates the line and column number for the specified offset into the buffer.
   *
   * @remarks
   * This is a potentially expensive operation.
   *
   * @param index - an integer offset
   * @param buffer - the buffer
   */
  public getLocation(index: number): ITextLocation {
    if (index < 0 || index > this.buffer.length) {
      // No match
      return { line: 0, column: 0 };
    }

    // TODO: Consider caching or optimizing this somehow
    let line: number = 1;
    let column: number = 1;

    let currentIndex: number = 0;

    while (currentIndex < index) {
      const current: string = this.buffer[currentIndex];
      ++currentIndex;

      if (current === '\r') {
        // CR
        // Ignore '\r' and assume it will always have an accompanying '\n'
        continue;
      }

      if (current === '\n') {
        // LF
        ++line;
        column = 1;
      } else {
        // NOTE: For consistency with the TypeScript compiler, a tab character is assumed
        // to advance by one column
        ++column;
      }
    }

    return { line, column };
  }

  private _validateBounds(): TextRange {
    if (this.pos < 0) {
      throw new Error('TextRange.pos cannot be negative');
    }
    if (this.end < 0) {
      throw new Error('TextRange.end cannot be negative');
    }
    if (this.end < this.pos) {
      throw new Error('TextRange.end cannot be smaller than TextRange.pos');
    }
    if (this.pos > this.buffer.length) {
      throw new Error('TextRange.pos cannot exceed the associated text buffer length');
    }
    if (this.end > this.buffer.length) {
      throw new Error('TextRange.end cannot exceed the associated text buffer length');
    }

    return this;
  }
}
