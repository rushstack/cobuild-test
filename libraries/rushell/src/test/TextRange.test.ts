// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { TextRange } from '../TextRange';

function escape(s: string): string {
  return s.replace(/\n/g, '[n]').replace(/\r/g, '[r]').replace(/\t/g, '[t]');
}

function matchSnapshot(textRange: TextRange): void {
  for (let i: number = -1; i <= textRange.end + 1; ++i) {
    // Show the current character
    const c: string = escape(textRange.buffer.substr(Math.max(i, 0), 1))
      .replace(/\n/g, '[n]')
      .replace(/\r/g, '[r]');

    // Show the next 10 characters of context
    const context: string = escape(textRange.buffer.substr(Math.max(i, 0), 10));

    expect({
      c: c,
      context: context,
      i: i,
      location: textRange.getLocation(i)
    }).toMatchSnapshot();
  }
}

test('construction scenarios', () => {
  const buffer: string = '0123456789';
  const textRange: TextRange = TextRange.fromString(buffer);
  expect(textRange.toString()).toEqual(buffer);

  const subRange: TextRange = textRange.getNewRange(3, 6);
  expect(subRange).toMatchSnapshot('subRange');
});

test('getLocation() basic', () => {
  const textRange: TextRange = TextRange.fromString(
    [
      'L1',
      'L2',
      '', // (line 3 is blank)
      'L4',
      'L5+CR\rL5+CRLF\r\nL6+LFCR\n\rL7'
    ].join('\n')
  );
  matchSnapshot(textRange);
});

test('getLocation() empty string', () => {
  const textRange: TextRange = TextRange.fromString('');
  matchSnapshot(textRange);
});

test('getLocation() CR string', () => {
  const textRange: TextRange = TextRange.fromString('\r');
  matchSnapshot(textRange);
});

test('getLocation() LF string', () => {
  const textRange: TextRange = TextRange.fromString('\n');
  matchSnapshot(textRange);
});

test('getLocation() tab characters', () => {
  // Tab character advances by only one column
  const textRange: TextRange = TextRange.fromString('1\t3');
  matchSnapshot(textRange);
});
