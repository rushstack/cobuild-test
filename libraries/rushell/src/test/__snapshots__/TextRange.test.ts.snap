// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`construction scenarios: subRange 1`] = `
TextRange {
  "buffer": "0123456789",
  "end": 6,
  "pos": 3,
}
`;

exports[`getLocation() CR string 1`] = `
Object {
  "c": "[r]",
  "context": "[r]",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() CR string 2`] = `
Object {
  "c": "[r]",
  "context": "[r]",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`getLocation() CR string 3`] = `
Object {
  "c": "",
  "context": "",
  "i": 1,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`getLocation() CR string 4`] = `
Object {
  "c": "",
  "context": "",
  "i": 2,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() LF string 1`] = `
Object {
  "c": "[n]",
  "context": "[n]",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() LF string 2`] = `
Object {
  "c": "[n]",
  "context": "[n]",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`getLocation() LF string 3`] = `
Object {
  "c": "",
  "context": "",
  "i": 1,
  "location": Object {
    "column": 1,
    "line": 2,
  },
}
`;

exports[`getLocation() LF string 4`] = `
Object {
  "c": "",
  "context": "",
  "i": 2,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() basic 1`] = `
Object {
  "c": "L",
  "context": "L1[n]L2[n][n]L4[n]",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() basic 2`] = `
Object {
  "c": "L",
  "context": "L1[n]L2[n][n]L4[n]",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`getLocation() basic 3`] = `
Object {
  "c": "1",
  "context": "1[n]L2[n][n]L4[n]L",
  "i": 1,
  "location": Object {
    "column": 2,
    "line": 1,
  },
}
`;

exports[`getLocation() basic 4`] = `
Object {
  "c": "[n]",
  "context": "[n]L2[n][n]L4[n]L5",
  "i": 2,
  "location": Object {
    "column": 3,
    "line": 1,
  },
}
`;

exports[`getLocation() basic 5`] = `
Object {
  "c": "L",
  "context": "L2[n][n]L4[n]L5+",
  "i": 3,
  "location": Object {
    "column": 1,
    "line": 2,
  },
}
`;

exports[`getLocation() basic 6`] = `
Object {
  "c": "2",
  "context": "2[n][n]L4[n]L5+C",
  "i": 4,
  "location": Object {
    "column": 2,
    "line": 2,
  },
}
`;

exports[`getLocation() basic 7`] = `
Object {
  "c": "[n]",
  "context": "[n][n]L4[n]L5+CR",
  "i": 5,
  "location": Object {
    "column": 3,
    "line": 2,
  },
}
`;

exports[`getLocation() basic 8`] = `
Object {
  "c": "[n]",
  "context": "[n]L4[n]L5+CR[r]",
  "i": 6,
  "location": Object {
    "column": 1,
    "line": 3,
  },
}
`;

exports[`getLocation() basic 9`] = `
Object {
  "c": "L",
  "context": "L4[n]L5+CR[r]L",
  "i": 7,
  "location": Object {
    "column": 1,
    "line": 4,
  },
}
`;

exports[`getLocation() basic 10`] = `
Object {
  "c": "4",
  "context": "4[n]L5+CR[r]L5",
  "i": 8,
  "location": Object {
    "column": 2,
    "line": 4,
  },
}
`;

exports[`getLocation() basic 11`] = `
Object {
  "c": "[n]",
  "context": "[n]L5+CR[r]L5+",
  "i": 9,
  "location": Object {
    "column": 3,
    "line": 4,
  },
}
`;

exports[`getLocation() basic 12`] = `
Object {
  "c": "L",
  "context": "L5+CR[r]L5+C",
  "i": 10,
  "location": Object {
    "column": 1,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 13`] = `
Object {
  "c": "5",
  "context": "5+CR[r]L5+CR",
  "i": 11,
  "location": Object {
    "column": 2,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 14`] = `
Object {
  "c": "+",
  "context": "+CR[r]L5+CRL",
  "i": 12,
  "location": Object {
    "column": 3,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 15`] = `
Object {
  "c": "C",
  "context": "CR[r]L5+CRLF",
  "i": 13,
  "location": Object {
    "column": 4,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 16`] = `
Object {
  "c": "R",
  "context": "R[r]L5+CRLF[r]",
  "i": 14,
  "location": Object {
    "column": 5,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 17`] = `
Object {
  "c": "[r]",
  "context": "[r]L5+CRLF[r][n]",
  "i": 15,
  "location": Object {
    "column": 6,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 18`] = `
Object {
  "c": "L",
  "context": "L5+CRLF[r][n]L",
  "i": 16,
  "location": Object {
    "column": 6,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 19`] = `
Object {
  "c": "5",
  "context": "5+CRLF[r][n]L6",
  "i": 17,
  "location": Object {
    "column": 7,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 20`] = `
Object {
  "c": "+",
  "context": "+CRLF[r][n]L6+",
  "i": 18,
  "location": Object {
    "column": 8,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 21`] = `
Object {
  "c": "C",
  "context": "CRLF[r][n]L6+L",
  "i": 19,
  "location": Object {
    "column": 9,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 22`] = `
Object {
  "c": "R",
  "context": "RLF[r][n]L6+LF",
  "i": 20,
  "location": Object {
    "column": 10,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 23`] = `
Object {
  "c": "L",
  "context": "LF[r][n]L6+LFC",
  "i": 21,
  "location": Object {
    "column": 11,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 24`] = `
Object {
  "c": "F",
  "context": "F[r][n]L6+LFCR",
  "i": 22,
  "location": Object {
    "column": 12,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 25`] = `
Object {
  "c": "[r]",
  "context": "[r][n]L6+LFCR[n]",
  "i": 23,
  "location": Object {
    "column": 13,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 26`] = `
Object {
  "c": "[n]",
  "context": "[n]L6+LFCR[n][r]",
  "i": 24,
  "location": Object {
    "column": 13,
    "line": 5,
  },
}
`;

exports[`getLocation() basic 27`] = `
Object {
  "c": "L",
  "context": "L6+LFCR[n][r]L",
  "i": 25,
  "location": Object {
    "column": 1,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 28`] = `
Object {
  "c": "6",
  "context": "6+LFCR[n][r]L7",
  "i": 26,
  "location": Object {
    "column": 2,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 29`] = `
Object {
  "c": "+",
  "context": "+LFCR[n][r]L7",
  "i": 27,
  "location": Object {
    "column": 3,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 30`] = `
Object {
  "c": "L",
  "context": "LFCR[n][r]L7",
  "i": 28,
  "location": Object {
    "column": 4,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 31`] = `
Object {
  "c": "F",
  "context": "FCR[n][r]L7",
  "i": 29,
  "location": Object {
    "column": 5,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 32`] = `
Object {
  "c": "C",
  "context": "CR[n][r]L7",
  "i": 30,
  "location": Object {
    "column": 6,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 33`] = `
Object {
  "c": "R",
  "context": "R[n][r]L7",
  "i": 31,
  "location": Object {
    "column": 7,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 34`] = `
Object {
  "c": "[n]",
  "context": "[n][r]L7",
  "i": 32,
  "location": Object {
    "column": 8,
    "line": 6,
  },
}
`;

exports[`getLocation() basic 35`] = `
Object {
  "c": "[r]",
  "context": "[r]L7",
  "i": 33,
  "location": Object {
    "column": 1,
    "line": 7,
  },
}
`;

exports[`getLocation() basic 36`] = `
Object {
  "c": "L",
  "context": "L7",
  "i": 34,
  "location": Object {
    "column": 1,
    "line": 7,
  },
}
`;

exports[`getLocation() basic 37`] = `
Object {
  "c": "7",
  "context": "7",
  "i": 35,
  "location": Object {
    "column": 2,
    "line": 7,
  },
}
`;

exports[`getLocation() basic 38`] = `
Object {
  "c": "",
  "context": "",
  "i": 36,
  "location": Object {
    "column": 3,
    "line": 7,
  },
}
`;

exports[`getLocation() basic 39`] = `
Object {
  "c": "",
  "context": "",
  "i": 37,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() empty string 1`] = `
Object {
  "c": "",
  "context": "",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() empty string 2`] = `
Object {
  "c": "",
  "context": "",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`getLocation() empty string 3`] = `
Object {
  "c": "",
  "context": "",
  "i": 1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() tab characters 1`] = `
Object {
  "c": "1",
  "context": "1[t]3",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`getLocation() tab characters 2`] = `
Object {
  "c": "1",
  "context": "1[t]3",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`getLocation() tab characters 3`] = `
Object {
  "c": "[t]",
  "context": "[t]3",
  "i": 1,
  "location": Object {
    "column": 2,
    "line": 1,
  },
}
`;

exports[`getLocation() tab characters 4`] = `
Object {
  "c": "3",
  "context": "3",
  "i": 2,
  "location": Object {
    "column": 3,
    "line": 1,
  },
}
`;

exports[`getLocation() tab characters 5`] = `
Object {
  "c": "",
  "context": "",
  "i": 3,
  "location": Object {
    "column": 4,
    "line": 1,
  },
}
`;

exports[`getLocation() tab characters 6`] = `
Object {
  "c": "",
  "context": "",
  "i": 4,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;
